File: hanoi.c
#include <stdio.h>
#include <stdlib.h>

int nbMove = 0;

void move(int n, int from, int via, int to) {
	if (n > 1) {
		nbMove++;
		move(n - 1, from, to, via);
		move(n - 1, via, from, to);
	}
}
int main(int argc, char *argv[]) {
	if(argc == 1) {
		printf("\tUtilisation : ./hanoi nb_disques, par exemple, ./hanoi 5\n");
		exit(13);
	}
	int nbDisks = strtol(argv[1], NULL, 10);
	move(nbDisks, 1,2,3);
	printf("%d moves are required!\n", nbMove);
	return 0;
}

File: Hanoi.java
public class Hanoi {
	private int nbMove = 0;

	public void move(int n, int from, int to, int via) {
		if (n > 1) {
			nbMove++;
			move(n - 1, from, to, via);
			move(n - 1, via, from, to);
		}
	}

	public int getNbMove() {
		return nbMove;
	}

	public static void main(String[] args) {
		int nbDisks = Integer.parseInt(args[0]);
		Hanoi h = new Hanoi();
		h.move(nbDisks, 1, 2, 3);
		System.out.println("Nb move: " + h.getNbMove());
	}
}
